### ISSUE TITLE
Connection Issues When Using Cellular

### ISSUE THREAD

**Issue by timali** [Azure/azure-iot-sdk-csharp#1720]
# Context

- OS, version, SKU and CPU architecture used: Windows 10 Desktop x64 and Windows 10 IoT Core ARM32
- Application's .NET Target Framework : UAP versions 1803 and 1809, and .NET Framework 4.7.2
- Device: Windows 10 desktop and Intrinsyc Open-Q 410c System-on-Module
- SDK version used: C# SDK version 1.33.1

## Description of the issue
When the machine is connected to the Internet via a cellular modem, establishing a connection to the IoT hub fails intermittently. Sometimes it connects right away, other times the call to OpenAsync() takes many seconds, and eventually the status changed callback will be called with state=Disconnected_Retrying and reason=Communication_Error. After this, sometimes it will then connect, other times it will not.

The cellular modem connection itself is working fine and has a very strong signal. We have reproduced this at various sites across the world, and with many different pieces of hardware and cellular modems. The issue seems to only occur when connected via cellular, as everything works fine when connected via Ethernet or Wi-FI.

## Code sample exhibiting the issue
using System;
using Microsoft.Azure.Devices.Client;

namespace DeviceClientCellularIssue
{
    class Program
    {
        static void OnConnStatusChanged(ConnectionStatus status, ConnectionStatusChangeReason reason)
        {
            Console.WriteLine($">>CONNECTION_STATUS>> STATE: {status}, REASON: {reason}.");
        }

        static void Main(string[] args)
        {
            DeviceClient client = DeviceClient.CreateFromConnectionString(<CONNECTION_STRING>);
            client.SetConnectionStatusChangesHandler(OnConnStatusChanged);

            Console.WriteLine("Opening the connection...");
            client.OpenAsync().Wait();

            Console.WriteLine("Connection opened. Press enter to continue...");
            Console.ReadLine();
        }
    }
}

## Example Output of Failing Case:

c:\ConsoleApp1\bin\Debug>ConsoleApp1.exe
Opening the connection...
(About a 30 second delay here)
>>CONNECTION_STATUS>> STATE: Disconnected_Retrying, REASON: Communication_Error.
>>CONNECTION_STATUS>> STATE: Connected, REASON: Connection_Ok.
Connection opened. Press enter to continue...

## Console log of the issue
I tried setting the DEBUG environment variable to '*', but I did not see any additional console prints. I would be happy to reproduce again with instructions on how to enable the debug prints.

**Comment by abhipsaMisra** [comment]
You'll need to follow the instructions listed [here](https://github.com/Azure/azure-iot-sdk-csharp/tree/master/tools/CaptureLogs), to capture the sdk logs.
Our SDK logs are written to `EventSource`, so you should be able to use any tool capable of capturing it on Win IoT Core.
Afaik, `tracelog` command is available on Win IoT Core, but you'll need to send the list of providers in a format compatible with the tracing tool.

**Comment by timali** [comment]
[Reproduced_with_Logging_Enabled.zip](https://github.com/Azure/azure-iot-sdk-csharp/files/5732210/Reproduced_with_Logging_Enabled.zip)


I reproduced the issue with logging enabled. In the failing case (using cellular), it hangs on this line:

> 2020-12-22T17:07:49.9491673 [Microsoft-Azure-Devices-Device-Client-Info] (TlsTransport#40535505,  source:tls0. operation:.Execute. detail:.BeginOpen, AmqpLogOperationVerbose:).

for one minute, and then it times out, and eventually connects after about two minutes in total.

For reference, I also included a working case when using Ethernet, in which it connects immediately without issue.

**Comment by abhipsaMisra** [comment]
Thanks for sharing the sdk logs. 

This is the series of operations that are being carried out when connecting over the cellular network:
- Initialize a client instance to communicate over Amqp (tcp, with websockets as fallback).
- The sdk tries to establish TLS over TCP connection (port 5671), and fails (which you've correctly highlighted above):
```
[Microsoft-Azure-Devices-Device-Client-Info] (AmqpTransportInitiator#21210914,  source:tp-initiator. operation:.Connect. detail:.captium-iot-prod.azure-devices.net:5671, AmqpLogOperationInformational:).
[Microsoft-Azure-Devices-Device-Client-Info] (TlsTransport#40535505,  source:tls0. operation:.Execute. detail:.BeginOpen, AmqpLogOperationVerbose:).
[Microsoft-Azure-Devices-Device-Client-Info] (TlsTransport#40535505,  source:tls0. operation:.Execute. detail:.BeginClose, AmqpLogOperationVerbose:).
[Microsoft-Azure-Devices-Device-Client-Info] (TcpTransport#13574358,  source:tcp0. operation:.Execute. detail:.BeginClose, AmqpLogOperationVerbose:).
[Microsoft-Azure-Devices-Device-Client-Info] (TcpTransport#13574358,  source:tcp0. operation:.Execute. detail:.EndClose, AmqpLogOperationVerbose:).
[Microsoft-Azure-Devices-Device-Client-Info] (TlsTransport#40535505,  source:tls0. operation:.Execute. detail:.EndClose, AmqpLogOperationVerbose:).
```
- The sdk tries to establish the connection over secure websockets (port 443), and fails:
```
[Microsoft-Azure-Devices-Device-Client-Enter] (AmqpIoTTransport#22542438, CreateClientWebSocketTransportAsync, (00:01:00)).
[Microsoft-Azure-Devices-Device-Client-Enter] (AmqpIoTTransport#22542438, CreateClientWebSocketAsync, (00:01:00)).
[Microsoft-Azure-Devices-Device-Client-Exit] (AmqpIoTTransport#22542438, CreateClientWebSocketAsync, 00:01:00).
[Microsoft-Azure-Devices-Device-Client-Exit] (AmqpIoTTransport#22542438, CreateClientWebSocketTransportAsync, 00:01:00).

[Microsoft-Azure-Devices-Device-Client-ErrorMessage] (ErrorDelegatingHandler#4094363, ExecuteWithErrorHandlingAsync, Exception caught: System.Net.WebSockets.WebSocketException (2043875328): Unable to connect to the remote server ---> System.Exception: Timeout ---> System.Exception: The text associated with this error code could not be found.

The operation timed out

   at System.Net.WebSockets.WinRTWebSocket.ConnectAsync(Uri uri, CancellationToken cancellationToken, ClientWebSocketOptions options)
   at System.Net.WebSockets.WebSocketHandle.ConnectAsyncCore(Uri uri, CancellationToken cancellationToken, ClientWebSocketOptions options)
```
- The sdk tries over TCP again, and succeeds this time (_Without having root-caused this, I'd be hesitant to say that the 2nd attempt would always succeed_).
```
[Microsoft-Azure-Devices-Device-Client-Enter] (AmqpIoTConnector#9345130, OpenConnectionAsync, (00:01:00)).
[Microsoft-Azure-Devices-Device-Client-Enter] (AmqpIoTTransport#47606018, InitializeAsync, (00:01:00)).
[Microsoft-Azure-Devices-Device-Client-Info] (AmqpTransportInitiator#5689696,  source:tp-initiator. operation:.Connect. detail:.captium-iot-prod.azure-devices.net:5671, AmqpLogOperationInformational:).
[Microsoft-Azure-Devices-Device-Client-Info] (TlsTransport#8999294,  source:tls1. operation:.Execute. detail:.BeginOpen, AmqpLogOperationVerbose:).
[Microsoft-Azure-Devices-Device-Client-Info] (TlsTransport#8999294,  source:tls1. operation:.Execute. detail:.EndOpen, AmqpLogOperationVerbose:).
[Microsoft-Azure-Devices-Device-Client-Exit] (AmqpIoTTransport#47606018, InitializeAsync, 00:01:00).
[Microsoft-Azure-Devices-Device-Client-Info] (AmqpConnection#20222386,  source:out-connection1. operation:.Execute. detail:.BeginOpen, AmqpLogOperationVerbose:).
[Microsoft-Azure-Devices-Device-Client-Info] (AmqpConnection#20222386,  source:out-connection1. connection:.out-connection1, AmqpOpenConnection:).
[Microsoft-Azure-Devices-Device-Client-Info] (AmqpConnection#20222386,  source:out-connection1. operation:.S:HDR. fromState:.Start. toState:.HeaderSent, AmqpStateTransition:).
[Microsoft-Azure-Devices-Device-Client-Info] (AmqpConnection#20222386,  source:out-connection1. operation:.S:OPEN. fromState:.HeaderSent. toState:.OpenPipe, AmqpStateTransition:).
[Microsoft-Azure-Devices-Device-Client-Info] (AmqpConnection#20222386,  source:out-connection1. operation:.R:HDR. fromState:.OpenPipe. toState:.OpenSent, AmqpStateTransition:).
[Microsoft-Azure-Devices-Device-Client-Info] (AmqpConnection#20222386,  source:out-connection1. operation:.R:OPEN. fromState:.OpenSent. toState:.Opened, AmqpStateTransition:).
[Microsoft-Azure-Devices-Device-Client-Info] (AmqpConnection#20222386,  source:out-connection1. operation:.Execute. detail:.EndOpen, AmqpLogOperationVerbose:).
[Microsoft-Azure-Devices-Device-Client-Exit] (AmqpIoTConnector#9345130, OpenConnectionAsync, 00:01:00)
```

This reads to me as an underlying tcp/ socket related issue. 
- Could you check that there are no firewall/ proxy /outbound connection settings on your cellular network that are restricting these socket connections from being established?
- Another thing you could try for debugging is to establish a simple connection over Http, and see if that succeed.
- If none of the above steps give any insight, could you capture the network traces for the cellular connection attempt and share those with us?

**Comment by timali** [comment]
I will reproduce and capture a network trace. Will a Wireshark capture (.pcap) be sufficient, or is there some other method you would like for me to use to capture a trace? Also, if Wireshark is ok, would it be enough to capture only AMQP traffic on TCP port 5671, or would you prefer a capture of all traffic on the machine?

**Comment by abhipsaMisra** [comment]
A wireshark trace should be enough. If possible, could you capture all traffic on the machine? That'll help us identify if there is some sort of gateway involved, and at which point the connection times out.

**Comment by abhipsaMisra** [comment]
@timali - were you able to verify that there are no other firewall/ outbound connection configurations restricting socket connections from being established? Did you try establishing a connection with Https?

**Comment by timali** [comment]
I verified that there are no outgoing firewall or restrictions on the cellular network (there are no limitations at all on outgoing traffic). This makes sense because the connection issue over cellular is sporadic -- it works fine sometimes, but then other times it will hang upon attempting to connect. For example, during the last test I performed, it connected fine 7 times in a row, and only hung during the connection attempt the 8th time I tried it, all during the same power cycle and all tests performed within 5 minutes of each other. If the issue were with an outgoing firewall, I would think the issue would be persistent, and not sporadic.

Also, I can in general make HTTPS connections from the machines over the cellular interface just fine. I tested this by using Chrome to navigate to different websites using HTTPS, and each time, it connected just fine and without any delay. In fact, when I was testing the connection using the Azure Client SDK, during the 8th test where the Azure Client SDK hung for about 1 minute trying to connect, I successfully made several HTTPS connections to google.com using Chrome - and this was during the time the Azure Client SDK was apparently hung trying to make a TLS connection to Azure. This seems to indicate that the cellular interface was working fine, and that TLS in general was also working fine.

I captured a Wireshark trace of the success case, but I was unable to capture a trace of the failing case because apparently the packet capture driver that Wireshark uses is not compatible with the cellular modem driver. Whenever I try to capture packets from the cellular interface, Wireshark complains that the capture session can not be initiated. I've tried both promiscuous mode and non-promiscuous mode. So unless we can figure out another way to capture packets, I will be unable to capture a trace using the cellular interface.

I have attached what information I could collect: 1) A successful connection using Ethernet and corresponding packet capture, 2) A successful connection using cellular (no packet capture), and 3) a failed connection using cellular (no packet capture, but I was able to establish several HTTPS connections to google.com using Chrome while the Azure Client SDK was hung).

[Captures with WireShark.zip](https://github.com/Azure/azure-iot-sdk-csharp/files/5796479/Captures.with.WireShark.zip)

**Comment by abhipsaMisra** [comment]
Thank you for the very detailed analysis @timali . 
I would have to agree with your assessment, had this been a firewall issue, I would also expect it to affect every single outbound connection every time, and not sporadically.

The failure in the sdk happens at a point where we are using the Amqp library underneath to negotiate and open a TLS connection. Without any additional tcp stack related data, it is a bit difficult for me to put my finger on what exactly it is that is causing this issue. While I discuss this with my team, would it be possible for you to try the same for all below protocols (your current test uses tcp with a fallback to ws - I'd like to see if this is isolated to a tcp socket or a ws implementation):
- AMQP_tcp only
- AMQP_ws only
- MQTT_tcp only

**Comment by timali** [comment]
[Protocol Testing.zip](https://github.com/Azure/azure-iot-sdk-csharp/files/5876042/Protocol.Testing.zip)
I performed 10 tests with each protocol and recorded the results, along with the log for any failures. There is a summary of the results called "Readme.txt" in the archive. But, as a quick summary, both AMQP_TCP_Only and MQTT_TCP_Only showed similar failure rates (1 or 2 failures in 10 tests), while AMQP_Websocket_Only failed all 10 times. Ethernet passed 20/20 tests.

**Comment by jamdavi** [comment]
Hey @timali I am working weith @abhipsaMisra on this. When looking at the extensive logging and testing you've done it's apparent that WebSockets are not behaving well in conjunction with the cellular modem.

> So unless we can figure out another way to capture packets, I will be unable to capture a trace using the cellular interface.

We can try to use the NDIS ETL, I don't have a device with a cellular modem so I can't be 100% on this. Here are the commands you should run on the device with the modem.

```
New-NetEventSession -Name "Capture" -CaptureMode SaveToFile -LocalFilePath "c:\mytrace.etl"
Add-NetEventPacketCaptureProvider -SessionName "Capture" -Level 4 -CaptureType BothPhysicalAndSwitch
Start-NetEventSession -Name "Capture" -TruncationLength 1600

<<< do your testing>>>

Stop-NetEventSession -Name "Capture"
Remove-NetEventSession -Name "Capture"
```

Download [ETL to PCAPNG](https://github.com/microsoft/etl2pcapng/releases/tag/v1.4.0), extract it to where you have the capture file and run the following command.

```
etl2pcapng.exe mytrace.etl mytrace.pcap
```



[New-NetEventSession](https://docs.microsoft.com/en-us/powershell/module/neteventpacketcapture/new-neteventsession?view=win10-ps)
[Add-NetEventPacketCaptureProvider](https://docs.microsoft.com/en-us/powershell/module/neteventpacketcapture/add-neteventpacketcaptureprovider?view=win10-ps)
[Start-NetEventSession](https://docs.microsoft.com/en-us/powershell/module/neteventpacketcapture/start-neteventsession?view=win10-ps)
[Stop-NetEventSession](https://docs.microsoft.com/en-us/powershell/module/neteventpacketcapture/stop-neteventsession?view=win10-ps)
[Remove-NetEventSession](https://docs.microsoft.com/en-us/powershell/module/neteventpacketcapture/remove-neteventsession?view=win10-ps)

**Comment by jamdavi** [comment]
Hey @timali just wanted to check in on this to see if you had a chance to get the logs.

**Comment by abhipsaMisra** [comment]
@timali Were you able to collect the logs that @jamdavi had requested above?

**Comment by timali** [comment]
Apologies for the delay -- production issues and unrelated bugs took priority. But, I was able to reproduce the issue, and I was able to capture packet traces using the instructions provided for NDIS ETL.

[Captures with ETL.zip](https://github.com/Azure/azure-iot-sdk-csharp/files/5992221/Captures.with.ETL.zip)

There are two sets of traces: one using AMQP with WebSocket only, and one using AMQP with TCP only. Personally, I am more interested in TCP only since our application does not require WebSockets.

Within each test case, there is a log text file, a readme.txt describing the test case, the trace.etl, and the generate PCAP file.

The test case most interesting to me is "AMQP_TCP_Only\Cellular\Test 2 (Failure)". The Azure Client SDK hangs for one minute during the initial TLS connection attempt, and during this time, I was able to use Chrome to create a TLS connection to www.google.com and www.amazon.com to prove the cellular modem and TLS connections were working. After one minute, the initial connection attempt was aborted, and the Azure Client SDK was able to connect on a retry.

**Comment by jamdavi** [comment]
Thanks @timali. I also zeroed in on the TCP only connection. Below is what is different between a passed and failed connection. You seem to have some knowledge in this area so I'm just going to skip to the good stuff...

**TL;DR Packets are getting dropped before they reach the IoT Hub Client (and in all likely hood before it reaches the device). This is not something specific to the IoT Hub SDK and I don't think there is much troubleshooting I can do without spending many hours with a physical device.** 

## Next Steps
To me it seems like there is something local to the device or configuration of the physical hardware that could be causing this.

1. Review and tune network configuration settings; personally I'd start by lowering the MTU, start with 1400 and move down in 40 byte increments (right now I think you're at 1418)
2. Disable the windows firewall
3. Disable third party anti-virus and firewalls
4. If possible try the modem on another machine, see if the connection issue follows
5. Contact Microsoft Support for help with Windows 10 IoT Core
6. Contact your cell provider or the Modem manufacturer to troubleshoot dropped packets

## TLS Truncation

**In short what it looks like is something is truncating the server side TLS server records before they reach the network layer**. I see this specific behavior for three connections which are unrelated and made from different applications. One to IoT Hub (included), one to a telemetry service, and one for Acronis Cyber Cloud (included). I can't say WHAT this something is but I can say that it is only happening when you're on your cellular modem.

If you look at your connection to the IoT hub in both the failed trace and the passed trace you'll see that after the Client Hello (frame 14 in the passed image and frame 19 in the failed image) the IoT Hub has sent some data but the packet capture says the previous segments weren't captured. 

What is happening here is the first two segments have been removed by "the something" I mention above. The image below is from the good trace so it indicates that this is happening all the time. I know it's two segments because the next sequence number is 2777. The calculation for this is based on your MSS. Your MSS is 1388 and there are 2776 bytes not in the stream which is evident by the fact the Sequence number for frame (frame 16 in the good and 20 in the bad) is 2777 (2776+1).

When the truncation happens Windows ACKs the SYN-ACK again because it's realized something bad happened. In a good scenario this ACK causes the IoT Hub service to resend all of the header data and in a bad scenario it would appear that IoT Hub never receives the duplicate ACK and after 30 seconds it shuts down the connection.

### Passed Trace IoT Hub
![image](https://user-images.githubusercontent.com/73593426/108423822-5a94bf00-7206-11eb-80fb-dab0b982d5be.png)

### Failed Trace IoT Hub
![image](https://user-images.githubusercontent.com/73593426/108425228-2c17e380-7208-11eb-9354-09320554e886.png)

### Failed Trace Acronis
![image](https://user-images.githubusercontent.com/73593426/108529548-91b8ae00-72a2-11eb-84c4-fd597d5ffd7a.png)

## Three-way Handshake Weirdness
Aside from the TLS truncation I do see signs of the duplicate ACK happening during normal TCP streams. For example the picture below is one from Google. You can see what looks like a well established connection but your device still asks for the data over again. I'm not sure what would cause this other than the packets being malformed or missing.

![image](https://user-images.githubusercontent.com/73593426/108535387-48b82800-72a9-11eb-8157-fcfe8aa0e063.png)

**Comment by jamdavi** [comment]
@timlai, did you get a chance to try out the next steps? I think we're at a point that we can close this issue as we've determined the issue isn't solely with the SDK, nor is it with a single server.

**Comment by jamdavi** [comment]
Hey @timali, seems like I tagged a non existent account last week. Pinging again to see if you have checked this out. I'll set to close this issue at the end of the week if I haven't heard back.

**Comment by jamdavi** [comment]
Closing issue per previous investigation. If there is more evidence that this is specifically an IoT Hub or Client SDK issue we can reopen.
